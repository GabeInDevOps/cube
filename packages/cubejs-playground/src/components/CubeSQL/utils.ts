import { Query } from '@cubejs-client/core';

export function getMembersList(query: Query): string[] {
  const list: string[] = [
    ...(query.dimensions || []),
    ...(query.measures || []),
    ...(query.segments || []),
  ];

  query.timeDimensions?.forEach((td) => {
    list.push(td.dimension);
  });

  if (query.order) {
    let order: [string, string][] = [];

    if (!Array.isArray(query.order)) {
      order = Object.entries(query.order);
    } else {
      order = query.order;
    }

    order.forEach(([member]) => list.push(member));
  }

  return Array.from(new Set(list));
}
