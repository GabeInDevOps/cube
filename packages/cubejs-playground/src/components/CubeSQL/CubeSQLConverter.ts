import {
  BinaryFilter,
  Query,
  TQueryOrderArray,
  TQueryOrderObject,
  UnaryFilter,
} from '@cubejs-client/core';

import { getMembersList } from './utils';

enum MemberKind {
  Measure,
  Dimension,
  Segment,
  TimeDimension,
}

type ConverterMember = {
  kind: MemberKind;
  name: string;
  position?: number;
  granularity?: string;
};

export class CubeSQLConverter {
  protected members: ConverterMember[] = [];

  protected tables: string[] = [];

  public constructor(protected readonly query: Query) {
    this.decompose();
  }

  public convert() {
    return this.makeQuery();
  }

  protected decompose() {
    this.query.measures?.forEach((name) =>
      this.members.push({
        name,
        kind: MemberKind.Measure,
      })
    );

    this.query.dimensions?.forEach((name) =>
      this.members.push({
        name,
        kind: MemberKind.Dimension,
      })
    );

    this.query.segments?.forEach((name) =>
      this.members.push({
        name,
        kind: MemberKind.Segment,
      })
    );

    this.query.timeDimensions?.forEach((td) =>
      this.members.push({
        name: td.dimension,
        kind: MemberKind.TimeDimension,
        granularity: td.granularity,
      })
    );

    const pathsList = getMembersList(this.query);
    this.tables = Array.from(
      new Set(pathsList.map((path) => path.split('.')[0]))
    );
  }

  protected makeQuery() {
    const query: string[] = [];

    query.push('SELECT');
    const members = this.members.map((member, index) => {
      member.position = index;

      if (member.kind === MemberKind.Measure) {
        return this.makeMeasure(member.name);
      } else if (member.kind === MemberKind.Dimension) {
        return this.makeDimension(member.name);
      } else if (member.kind === MemberKind.TimeDimension) {
        return this.makeTimeDimension(member.name, member.granularity);
      }
    });
    query.push(members.join(', '));

    query.push(`FROM ${this.tables[0]}`);

    query.push(this.makeJoins());

    if (this.query.filters?.length) {
      const filters = this.query.filters.map((filter) =>
        this.makeFilter(filter as any)
      );

      if (filters) {
        query.push(`WHERE ${filters.join(' AND ')}`);
      }
    }

    const groupBy = this.members
      .filter((member) => member.kind !== MemberKind.Measure)
      .map((member) => {
        return member.name;
      });

    if (groupBy) {
      query.push(`GROUP BY ${groupBy.join(', ')}`);
    }

    if (this.query.order) {
      query.push(this.makeOrderBy(this.query.order));
    }

    if (this.query.limit) {
      query.push(`LIMIT ${this.query.limit}`);
    }

    return query.join(' ');
  }

  protected makeJoins() {
    if (this.tables.length > 1) {
      // Naive approach
      const [main, ...otherTables] = this.tables;

      return otherTables
        .map((tableName) => {
          return `LEFT JOIN ${tableName} ON ${main}.__cubeJoinField = ${tableName}.__cubeJoinField`;
        })
        .join(' ');
    }

    return '';
  }

  protected makeMeasure(name: string) {
    return `MEASURE(${name})`;
  }

  // todo: support boolean filters
  protected makeFilter(filter: BinaryFilter | UnaryFilter) {
    if (filter.operator === 'inDateRange') {
      return [
        filter.member,
        '>',
        this.escapeValue(filter.values[0]),
        'AND',
        filter.member,
        '<',
        this.escapeValue(filter.values[1]),
      ].join(' ');
    }

    if (['equals', 'notEquals'].includes(filter.operator) && filter.values) {
      const hasSingleValue = filter.values.length <= 1;
      const operator = {
        equals: hasSingleValue ? '=' : 'IN',
        notEquals: hasSingleValue ? '!=' : 'NOT IN',
      };

      const nullValue =
        filter.operator === 'notEquals' ? `OR ${filter.member} IS NULL` : '';

      const value = hasSingleValue
        ? this.escapeValue(filter.values[0])
        : `(${filter.values.map(this.escapeValue).join(', ')})`;

      const string = [
        filter.member,
        operator[filter.operator],
        value,
        nullValue,
      ]
        .join(' ')
        .trim();

      return `(${string})`;
    }

    return '';
  }

  protected makeDimension(name: string) {
    return name;
  }

  protected makeTimeDimension(name: string, granularity?: string) {
    if (!granularity) {
      return;
    }

    return `DATE_TRUNC('${granularity}', ${name})`;
  }

  protected escapeValue(value: string) {
    return `'${value}'`;
  }

  protected makeOrderBy(queryOrder: TQueryOrderObject | TQueryOrderArray) {
    const orderArray = Array.isArray(queryOrder)
      ? queryOrder
      : Object.entries(queryOrder);

    return `ORDER BY ${orderArray
      .map(([member, order]) => `${member} ${order.toUpperCase()}`)
      .join(', ')
      .trim()}`;
  }
}
