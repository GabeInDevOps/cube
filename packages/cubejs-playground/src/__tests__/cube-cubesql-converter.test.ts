import { Query } from '@cubejs-client/core';
import { CubeSQLConverter } from '../components/CubeSQL/CubeSQLConverter';

const queries: Query[] = [
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status', 'Users.state', 'Orders.createdAt'],
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status', 'Users.state', 'Orders.createdAt'],
    order: {
      'Orders.count': 'desc',
      'Orders.status': 'asc',
      'Users.state': 'desc',
    },
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status', 'Users.state', 'Orders.createdAt'],
    order: [
      ['Orders.count', 'desc'],
      ['Orders.status', 'asc'],
      ['Users.state', 'desc'],
    ],
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status'],
    timeDimensions: [
      {
        dimension: 'Orders.createdAt',
        granularity: 'day',
        dateRange: ['2020-01-01', '2021-01-01'],
      },
    ],
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status'],
    timezone: 'America/Los_Angeles',
    limit: 100,
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status'],
    timezone: 'America/Los_Angeles',
    limit: 100,
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status', 'Users.state'],
    filters: [
      {
        member: 'Users.state',
        operator: 'equals',
        values: ['US'],
      },
      {
        member: 'Orders.status',
        operator: 'equals',
        values: ['canceled', 'active'],
      },
    ],
  },
  {
    measures: ['Orders.count'],
    dimensions: ['Orders.status', 'Users.state'],
    filters: [
      {
        member: 'Users.state',
        operator: 'equals',
        values: ['US'],
      },
      {
        member: 'Users.state',
        operator: 'equals',
        values: ['Canada'],
      },
    ],
  },
  {
    measures: ['Orders.count'],
    timeDimensions: [
      {
        dimension: 'Orders.createdAt',
        granularity: 'quarter',
      },
    ],
    order: [
      ['Orders.count', 'asc'],
      ['Users.state', 'asc'],
    ],
    dimensions: ['Users.state', 'Orders.createdAt'],
  },
];

const q: Query =   {
  measures: ['Orders.count'],
  timeDimensions: [
    {
      dimension: 'Orders.createdAt',
      granularity: 'quarter',
    },
  ],
  filters: [
    {
      member: 'Orders.createdAt',
      operator: 'inDateRange',
      values: ['2020-01-01', '2020-01-01']
    },
    {
      member: 'Orders.status',
      // operator: 'notEquals',
      operator: 'equals',
      // values: ['completed']
      values: ['completed', 'shipped']
    }
  ],
  order: [
    ['Orders.count', 'asc'],
    ['Users.state', 'asc'],
  ],
  dimensions: ['Users.state', 'Orders.createdAt'],
};

test('Cube CubeSQL converter', () => {
  queries.forEach((query) => {
    const converter = new CubeSQLConverter(query);
    
    expect(converter.convert()).toMatchSnapshot(JSON.stringify(query));
  });
});
