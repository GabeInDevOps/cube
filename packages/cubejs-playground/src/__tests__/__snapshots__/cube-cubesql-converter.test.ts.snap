// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status","Users.state","Orders.createdAt"],"order":[["Orders.count","desc"],["Orders.status","asc"],["Users.state","desc"]]} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, Users.state, Orders.createdAt FROM Orders LEFT JOIN Users ON Orders.__cubeJoinField = Users.__cubeJoinField GROUP BY Orders.status, Users.state, Orders.createdAt ORDER BY Orders.count DESC, Orders.status ASC, Users.state DESC"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status","Users.state","Orders.createdAt"],"order":{"Orders.count":"desc","Orders.status":"asc","Users.state":"desc"}} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, Users.state, Orders.createdAt FROM Orders LEFT JOIN Users ON Orders.__cubeJoinField = Users.__cubeJoinField GROUP BY Orders.status, Users.state, Orders.createdAt ORDER BY Orders.count DESC, Orders.status ASC, Users.state DESC"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status","Users.state","Orders.createdAt"]} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, Users.state, Orders.createdAt FROM Orders LEFT JOIN Users ON Orders.__cubeJoinField = Users.__cubeJoinField GROUP BY Orders.status, Users.state, Orders.createdAt"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status","Users.state"],"filters":[{"member":"Users.state","operator":"equals","values":["US"]},{"member":"Orders.status","operator":"equals","values":["canceled","active"]}]} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, Users.state FROM Orders LEFT JOIN Users ON Orders.__cubeJoinField = Users.__cubeJoinField WHERE (Users.state = 'US') AND (Orders.status IN ('canceled', 'active')) GROUP BY Orders.status, Users.state"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status","Users.state"],"filters":[{"member":"Users.state","operator":"equals","values":["US"]},{"member":"Users.state","operator":"equals","values":["Canada"]}]} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, Users.state FROM Orders LEFT JOIN Users ON Orders.__cubeJoinField = Users.__cubeJoinField WHERE (Users.state = 'US') AND (Users.state = 'Canada') GROUP BY Orders.status, Users.state"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status"],"timeDimensions":[{"dimension":"Orders.createdAt","granularity":"day","dateRange":["2020-01-01","2021-01-01"]}]} 1`] = `"SELECT MEASURE(Orders.count), Orders.status, DATE_TRUNC('day', Orders.createdAt) FROM Orders  GROUP BY Orders.status, Orders.createdAt"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status"],"timezone":"America/Los_Angeles","limit":100} 1`] = `"SELECT MEASURE(Orders.count), Orders.status FROM Orders  GROUP BY Orders.status LIMIT 100"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"dimensions":["Orders.status"],"timezone":"America/Los_Angeles","limit":100} 2`] = `"SELECT MEASURE(Orders.count), Orders.status FROM Orders  GROUP BY Orders.status LIMIT 100"`;

exports[`Cube CubeSQL converter: {"measures":["Orders.count"],"timeDimensions":[{"dimension":"Orders.createdAt","granularity":"quarter"}],"order":[["Orders.count","asc"],["Users.state","asc"]],"dimensions":["Users.state","Orders.createdAt"]} 1`] = `"SELECT MEASURE(Orders.count), Users.state, Orders.createdAt, DATE_TRUNC('quarter', Orders.createdAt) FROM Users LEFT JOIN Orders ON Users.__cubeJoinField = Orders.__cubeJoinField GROUP BY Users.state, Orders.createdAt, Orders.createdAt ORDER BY Orders.count ASC, Users.state ASC"`;
